### PR TITLE
Add compatibility layer for handling Cask Tap migration.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

### CODE REVIEW

File: Library/Homebrew/compat/tap.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+module CaskTapMigration
+  def initialize(user, repo)
+    super
+
+    return unless user == "caskroom"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/compat/tap.rb]: Make this an if?

File: Library/Homebrew/compat/tap.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    # TODO: Remove this check after migration.
+    return unless repo == "tap-migration-test"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/compat/tap.rb]: Make this an if?

File: Library/Homebrew/compat/tap.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    new_user = "Homebrew"
+    new_repo = (repo == "cask") ? repo : "cask-#{repo}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/compat/tap.rb]: If else?

File: Library/Homebrew/compat/tap.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    super(new_user, new_repo)
+
+    return unless old_path.git?
+
+    new_name = name
+    new_path = path
+    new_remote = default_remote
+
+    ohai "Migrating Tap #{old_name} to #{new_name} …"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/compat/tap.rb]: Lowercase t on tap and use ... rather than Unicode thing

File: Library/Homebrew/extend/git_repository.rb [hunk 5]
```diff
@@ -13,6 +13,13 @@ def git_origin
     end
   end
 
+  def git_origin=(origin)
+    return unless git? && Utils.git_available?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/git_repository.rb]: Make if rather than unless

File: Library/Homebrew/compat/tap.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    path.dirname.mkpath
+    FileUtils.mv old_path, new_path
+
+    puts "Changing remote from #{old_remote} to #{new_remote} …"
+    new_path.git_origin = new_remote
+  end
+end
+
+class Tap
+  prepend CaskTapMigration
```
Comment (MikeMcQuaid) [inline Library/Homebrew/compat/tap.rb]: This isn’t how we generally do compat classes; might be worth a PR to update the others to this format if it’s more desirable?

File: Library/Homebrew/missing_formula.rb [hunk 7]
```diff
@@ -101,7 +101,7 @@ def tap_migration_reason(name)
           EOS
           break if new_tap_name == CoreTap.instance.name
 
-          install_cmd = if new_tap_user == "caskroom"
+          install_cmd = if new_tap_user == "caskroom" || (new_tap_user == "homebrew" && new_tap_repo.start_with?("cask"))
```
Comment (MikeMcQuaid) [inline Library/Homebrew/missing_formula.rb]: Feels a bit much for one line

File: Library/Homebrew/tap.rb [hunk 8]
```diff
@@ -26,7 +26,7 @@ def self.fetch(*args)
     end
 
     # We special case homebrew and linuxbrew so that users don't have to shift in a terminal.
-    user = user.capitalize if ["homebrew", "linuxbrew"].include? user
+    user = user.capitalize if ["homebrew", "linuxbrew"].include?(user)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/tap.rb]: Any reason for this change?

File: Library/Homebrew/compat/tap.rb (hunk 2 above)
Comment (reitermarkus) [reply Library/Homebrew/compat/tap.rb]: How?

File: Library/Homebrew/compat/tap.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/compat/tap.rb]: How?

File: Library/Homebrew/missing_formula.rb (hunk 7 above)
Comment (reitermarkus) [reply Library/Homebrew/missing_formula.rb]: For what it's worth, this is only temporary.

**Comment by reitermarkus** [comment]
By the way, to test this, run:

```
brew tap caskroom/tap-migration-test --no-compat
brew untap caskroom/tap-migration-test
```

File: Library/Homebrew/compat/tap.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/compat/tap.rb]: `return if repo != "tap-migration-test"`

File: Library/Homebrew/extend/git_repository.rb (hunk 5 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/extend/git_repository.rb]: @reitermarkus Could update them too; it's pretty yuck.

File: Library/Homebrew/missing_formula.rb (hunk 7 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/missing_formula.rb]: How temporary, roughly?

File: Library/Homebrew/compat/tap.rb (hunk 2 above)
Comment (reitermarkus) [reply Library/Homebrew/compat/tap.rb]: I'm pretty sure RuboCop would complain about that.

File: Library/Homebrew/missing_formula.rb (hunk 7 above)
Comment (reitermarkus) [reply Library/Homebrew/missing_formula.rb]: It can be removed after the migration.

File: Library/Homebrew/extend/git_repository.rb (hunk 5 above)
Comment (reitermarkus) [reply Library/Homebrew/extend/git_repository.rb]: On another note, would we even get to a point where we call any of these when `Utils.git_available?` is `false`? I mean, how can there be a `.git` directory when `git` is not available?

File: Library/Homebrew/compat/tap.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/compat/tap.rb]: @reitermarkus Try it; pretty sure it doesn't 😉

File: Library/Homebrew/missing_formula.rb (hunk 7 above)
Comment (reitermarkus) [reply Library/Homebrew/missing_formula.rb]: Exactly.

Comment (JCount) [reply Library/Homebrew/missing_formula.rb]: I concur.

**Comment by MikeMcQuaid** [comment]
Thanks @reitermarkus!

**Comment by JCount** [comment]
🎊

**Pull request merged**